pkg/plugins/tainttoleration/taint_toleration.go

```diff
@@ -38,80 +44,150 @@ func (pl *TaintToleration) Description() string {
 	return description
 }
 
-func (pl *TaintToleration) Filter(ctx context.Context, placement *clusterapiv1beta1.Placement, clusters []*clusterapiv1.ManagedCluster) ([]*clusterapiv1.ManagedCluster, error) {
+func (pl *TaintToleration) Filter(ctx context.Context, placement *clusterapiv1beta1.Placement, clusters []*clusterapiv1.ManagedCluster) plugins.PluginFilterResult {
 	if len(clusters) == 0 {
-		return clusters, nil
+		return plugins.PluginFilterResult{
+			Filtered: clusters,
+		}
 	}
 
 	// do validation on each toleration and return error if necessary
 	for _, toleration := range placement.Spec.Tolerations {
 		if len(toleration.Key) == 0 && toleration.Operator != clusterapiv1beta1.TolerationOpExists {
-			return nil, errors.New("If the key is empty, operator must be Exists.\n")
+			return plugins.PluginFilterResult{
+				Err: errors.New("If the key is empty, operator must be Exists.\n"),
+			}
 		}
 		if toleration.Operator == clusterapiv1beta1.TolerationOpExists && len(toleration.Value) > 0 {
-			return nil, errors.New("If the operator is Exists, the value should be empty.\n")
+			return plugins.PluginFilterResult{
+				Err: errors.New("If the operator is Exists, the value should be empty.\n"),
+			}
 		}
 	}
 
-	existingDecisions := getDecisions(pl.handle, placement)
+	decisionClusterNames := getDecisionClusterNames(pl.handle, placement)
 
 	// filter the clusters
 	matched := []*clusterapiv1.ManagedCluster{}
 	for _, cluster := range clusters {
-		if isClusterTolerated(cluster, placement.Spec.Tolerations, existingDecisions.Has(cluster.Name)) {
+		if tolerated, _ := isClusterTolerated(cluster, placement.Spec.Tolerations, decisionClusterNames.Has(cluster.Name)); tolerated {
 			matched = append(matched, cluster)
 		}
 	}
 
-	return matched, nil
+	return plugins.PluginFilterResult{
+		Filtered: matched,
+	}
+}
+
+func (pl *TaintToleration) RequeueAfter(ctx context.Context, placement *clusterapiv1beta1.Placement) plugins.PluginRequeueResult {
+	// get exist decisions clusters
+	decisionClusterNames, decisionClusters := getDecisionClusters(pl.handle, placement)
+	if decisionClusterNames == nil || decisionClusters == nil {
+		return plugins.PluginRequeueResult{}
+	}
+
+	requeueList := []*plugins.PluginRequeueResult{}
+	// filter and record pluginRequeueResults
+	for _, cluster := range decisionClusters {
+		tolerated, requeue := isClusterTolerated(cluster, placement.Spec.Tolerations, decisionClusterNames.Has(cluster.Name))
+		if tolerated && requeue != nil {
+			requeueList = append(requeueList, requeue)
```

Comment:
nit: we do not need a separated sorting here. you could know the min requeue during the for loop. This could reduce some computation cost. But it can be a followup

Reply:
fixed.